octodns/provider/azuredns.py

```diff
@@ -0,0 +1,369 @@
+        client_id:
+        # Authentication Key Value req:
+        key:
+        # Directory ID (referred to tenant ID) req:
+        directory_id:
+        # Subscription ID req:
+        sub_id:
+        # Resource Group name req:
+        resource_group:
+
+        TODO: change config file to use env vars instead of hard-coded keys
+
+        personal notes: testing: test authentication vars located in
+                /home/t-hehwan/vars.txt
+    '''
+    SUPPORTS_GEO = False
+    SUPPORTS = set(('A', 'AAAA', 'CNAME', 'MX', 'NS', 'PTR', 'SRV', 'TXT'))
+
+    def __init__(self, id, client_id, key, directory_id, sub_id,
+                 resource_group, *args, **kwargs):
+        self.log = logging.getLogger('AzureProvider[{}]'.format(id))
+        self.log.debug('__init__: id=%s, client_id=%s, '
+                       'key=***, directory_id:%s', id, client_id, directory_id)
+        super(AzureProvider, self).__init__(id, *args, **kwargs)
+
+        credentials = ServicePrincipalCredentials(
+            client_id, secret=key, tenant=directory_id
+        )
+        self._dns_client = DnsManagementClient(credentials, sub_id)
+        self._resource_group = resource_group
+        self._azure_zones = set()
+
+    def _populate_zones(self):
+        self.log.debug('azure_zones: loading')
+        list_zones = self._dns_client.zones.list_by_resource_group
+        for zone in list_zones(self._resource_group):
+            self._azure_zones.add(zone.name)
+
+    def _check_zone(self, name, create=False):
+        '''
+            Checks whether a zone specified in a source exist in Azure server.
+            Note that Azure zones omit end '.' eg: contoso.com vs contoso.com.
+            Returns the name if it exists.
+
+            :param name: Name of a zone to checks
+            :type  name: str
+            :param create: If True, creates the zone of that name.
+            :type  create: bool
+
+            :type return: str or None
+        '''
+        self.log.debug('_check_zone: name=%s', name)
+        try:
+            if name in self._azure_zones:
+                return name
+            self._dns_client.zones.get(self._resource_group, name)
+            self._azure_zones.add(name)
+            return name
+        except:
+            if create:
+                self.log.debug('_check_zone:no matching zone; creating %s',
+                               name)
+                create_zone = self._dns_client.zones.create_or_update
+                create_zone(self._resource_group, name, Zone('global'))
+                return name
+            else:
+                raise
+
+    def populate(self, zone, target=False, lenient=False):
+        '''
+            Required function of manager.py.
+
+            Special notes for Azure.
+            Azure zone names omit final '.'
+            Azure root records names are represented by '@'. OctoDNS uses ''
+            Azure records created through online interface may have null values
```

Comment:
Interesting. That's case I haven't tested, but octoDNS might let you set that up with `values: []`. No clue what all the providers would do with it.